iocore/hostdb/P_RefCountCache.h

```diff
@@ -0,0 +1,896 @@
+    // found
+    this->metric_inc(refcountcache_total_hits_stat, 1);
+    return make_ptr((C *)l.m_value->item.get());
+  } else {
+    return Ptr<C>();
+  }
+}
+
+template <class C>
+void
+RefCountCachePartition<C>::put(uint64_t key, C *item, int size, int expire_time)
+{
+  this->metric_inc(refcountcache_total_inserts_stat, 1);
+  size += sizeof(C);
+  // Remove any colliding entries
+  this->erase(key);
+
+  // if we are full, and can't make space-- then don't store the item
+  if (this->is_full() && !this->reserve(size)) {
+    Debug("refcountcache", "partition %d is full-- not storing item key=%ld", this->part_num, key);
+    this->metric_inc(refcountcache_total_failed_inserts_stat, 1);
+    return;
+  }
+
+  // Create our value-- which has a ref to the `item`
+  RefCountCacheHashEntry *val = refCountCacheHashingValueAllocator.alloc();
+  val->set(item, key, size, expire_time);
+
+  // add expiry_entry to expiry queue, if the expire time is positive (otherwise it means don't expire)
+  if (expire_time >= 0) {
+    Debug("refcountcache", "partition %d adding entry with expire_time=%d\n", this->part_num, expire_time);
+    PriorityQueueEntry<RefCountCacheHashEntry *> *expiry_entry = expiryQueueEntry.alloc();
+    new ((void *)expiry_entry) PriorityQueueEntry<RefCountCacheHashEntry *>(val);
+    expiry_queue.push(expiry_entry);
+    val->expiry_entry = expiry_entry;
+  }
+
+  // add the item to the map
+  this->item_map.insert(val);
+  this->size += val->meta.size;
+  this->items++;
+  this->metric_inc(refcountcache_current_size_stat, (int64_t)val->meta.size);
+  this->metric_inc(refcountcache_current_items_stat, 1);
+}
+
+template <class C>
+void
+RefCountCachePartition<C>::erase(uint64_t key, int expiry_time)
+{
+  location_type l = this->item_map.find(key);
+  if (l.isValid()) {
+    if (expiry_time >= 0 && l.m_value->meta.expiry_time != expiry_time) {
+      return;
+    }
+    // found
+    this->item_map.remove(l);
+
+    // decrement usage counters
+    this->size -= l.m_value->meta.size;
+    this->items--;
+
+    this->metric_inc(refcountcache_current_size_stat, -((int64_t)l.m_value->meta.size));
+    this->metric_inc(refcountcache_current_items_stat, -1);
+
+    // remove from expiry queue
+    if (l.m_value->expiry_entry != NULL) {
+      Debug("refcountcache", "partition %d deleting item from expiry_queue idx=%d\n", this->part_num,
+            l.m_value->expiry_entry->index);
+      this->expiry_queue.erase(l.m_value->expiry_entry);
+      expiryQueueEntry.free(l.m_value->expiry_entry);
+      l.m_value->expiry_entry = NULL; // To avoid the destruction of `l` calling the destructor again-- and causing issues
+    }
+    // Since the Value is actually RefCountObj-- when this gets deleted normally it calls the wrong
+    // `free` method, this forces the delete/decr to happen with the right type
+    Ptr<C> *tmp = (Ptr<C> *)&l.m_value->item;
+    tmp->clear();
+    refCountCacheHashingValueAllocator.free(l.m_value);
```

Comment:
AFAICT the class allocator won't call any destructors, so you have to do that manually.
```C
l.m_value->~RefCountCacheHashEntry();
```
This should be equivalent.